### PR TITLE
Implement framework for observed hour angle-declination transforms.

### PR DESCRIPTION
**Pull request by mkbrewer**
### Description

This pull request is to address my request in issue #10239 for an observed Hour Angle/Declination frame to complement the AltAz frame. It's been a year now since I made this request and nothing has come of it, so I felt that it was necessary to take matters into my own hands. I have included a few basic tests here. Since the code is basically the same as that for AltAz, I don't think it is necessary to test for accuracy other than these. I did need to change the name of one existing test: `test_icrs_consistency()` to `test_icrs_altaz_consistency()`. I'm sure that there is likely to be some "glue" that I have neglected here and could use some help with that. I have made no attempt at amending the documentation as I wouldn't know where to begin with that. Also, I neglected to add [skip ci] to my commit message. Sorry about that.

### CODE REVIEW

File: astropy/coordinates/tests/test_intermediate_transformations.py
Comment (mhvk) [inline astropy/coordinates/tests/test_intermediate_transformations.py:170]: Is this statement true? I think we would need to add the relevant tests to `test_iau_fullstack`.

File: astropy/coordinates/builtin_frames/cirs_observed_transforms.py
Comment (mhvk) [inline astropy/coordinates/builtin_frames/cirs_observed_transforms.py:23]: In principle, we do not really need `coord_type`, since it is implicit in the frame (`coord_type = 'A' if insinstance(observed_frame, AltAz) else 'H'`). With that argument removed, the same function could serve both types of transforms directly. 

Though not super important.

File: astropy/coordinates/tests/test_intermediate_transformations.py
Comment (mhvk) [inline astropy/coordinates/tests/test_intermediate_transformations.py:183]: Type: should be `hadecframe`

**Review by mhvk — COMMENTED** [review]
This PR looks very good. I think the repeated tests probably cover most cases, but I wouldn't mind one or two simple tests that explicitly show that for some inputs, the HA and Dec are as expected (could be observing something near Spring equinox and showing that sources with RA~12 are at HA~0 at mightnight, and that their Dec is close.

Also, I think some documentation would be good - though I'm not 100% sure where it would be best.

**Comment by StuartLittlefair** [comment]
I just want to say that I will get round to reviewing this ASAP, but I am under a lot of work pressure until May 15th, so probably not before then.

File: astropy/coordinates/tests/test_intermediate_transformations.py
Comment (mkbrewer) [reply astropy/coordinates/tests/test_intermediate_transformations.py:170]: No, that was just a copy and paste from `test_cirs_to_altaz()`. I will remove it.

File: astropy/coordinates/tests/test_intermediate_transformations.py
Comment (mkbrewer) [reply astropy/coordinates/tests/test_intermediate_transformations.py:183]: Yep. Typo.

File: astropy/coordinates/builtin_frames/cirs_observed_transforms.py
Comment (mkbrewer) [reply astropy/coordinates/builtin_frames/cirs_observed_transforms.py:23]: I like this idea. Will do.

**Comment by mkbrewer** [comment]
Regarding tests for accuracy, my point is that these would be redundant. For instance, if `test_aa_high_precision()` passes, then it can be guaranteed that the corresponding test for `HADec` would also pass, should I create it. This is because both `AltAz` and `HADec` are derived from the same vector in `atioq()`.

File: astropy/coordinates/builtin_frames/cirs_observed_transforms.py
Comment (mhvk) [inline astropy/coordinates/builtin_frames/cirs_observed_transforms.py:90]: So, these functions are no longer needed; you can just do:
```
@frame_transform_graph.transform(FunctionTransformWithFiniteDifference, CIRS, AltAz)
@frame_transform_graph.transform(FunctionTransformWithFiniteDifference, CIRS, HADec)
def cirs_to_observed(cirs_coo, observed_frame):
    ...
```
and the same for `observed_to_cirs`.

Actually, maybe easiest if I just push directly to your branch, then I can do a rebase as well, so the tests can run.
def

Comment (mkbrewer) [reply astropy/coordinates/builtin_frames/cirs_observed_transforms.py:90]: Sure, go ahead.

**Comment by mhvk** [comment]
@mkbrewer - OK, I rebased to get rid of the merge conflicts and made the simplification I suggested. I also took the liberty to squash the multiple smaller commits. Locally, the tests pass, and on second thought I agree with you that there is not much point in separate precision tests. 

What still would be good, I think, is to have at least a simple example somewhere in the documentation, like
```
In [1]: from astropy import units as u

In [2]: from astropy.coordinates import EarthLocation, SkyCoord, HADec

In [3]: from astropy.time import Time

In [4]: el = EarthLocation.of_site('keck')

In [5]: t = Time('2010-03-21T10:00:00')

In [6]: sc = SkyCoord(12*u.hourangle, 10*u.deg)

In [7]: hadec = HADec(obstime=t, location=el)

In [8]: sc.transform_to(hadec)
Out[8]: 
<SkyCoord (HADec: obstime=2010-03-21T10:00:00.000, location=(-5464487.81759887, -2492806.59108569, 2151240.19451846) m, pressure=0.0 hPa, temperature=0.0 deg_C, relative_humidity=0.0, obswl=1.0 micron): (ha, dec) in deg
    (353.19976268, 9.93902734)>
```
(where I verified the result with the veritable `skycalc`)

File: astropy/coordinates/builtin_frames/hadec.py
Comment (mhvk) [inline astropy/coordinates/builtin_frames/hadec.py:91]: In the example I posted, I noticed that the units for `HA` are degrees, which is perhaps not what we want... One can change it by here using `RepresentationMapping('lon', 'ha', u.hourangle)`

Comment (mkbrewer) [reply astropy/coordinates/builtin_frames/hadec.py:91]: Yes, I noticed that also. Hourangle is definitely more appropriate here.

Comment (mkbrewer) [reply astropy/coordinates/builtin_frames/hadec.py:91]: Can I just pull from my branch to get your changes? Or does the rebase interfere with that somehow? I've never done a rebase myself.

Comment (mkbrewer) [reply astropy/coordinates/builtin_frames/hadec.py:91]: Thanks. That appears to have worked.

**Comment by mkbrewer** [comment]
Regarding documentation, can you handle that? I can't find any similar example for either `AltAz` or `TETE` anywhere in the `docs` directory, so I wouldn't know where to start.

**Comment by mhvk** [comment]
@mkbrewer - hmm, you're right, it is actually far from obvious where a simple example about how to use `HADec` (or `AltAz` for that matter) would go in the documentation. I do see that the list of frames available in, e.g., https://docs.astropy.org/en/latest/coordinates/skycoord.html is quite out of date. So, let me (1) cc @eteq, @adrn for possible suggestions; and (2) suggest that perhaps we should leave this to a separate issue that is maybe more specifically about the various frames and their uses (which I'll raise if the others agree).

I think otherwise this PR is in very good shape, though ideally @StuartLittlefair has a look as well (@StuartLittlefair - the addition is actually as simple as @mkbrewer had promised it would be).

File: docs/changes/coordinates/11676.feature.rst
Comment (pllim) [inline docs/changes/coordinates/11676.feature.rst:2]: Unless the user is supposed to run these tests, there is no need to mention them.
```suggestion
Adds new ``HADec`` built-in frame with transformations to/from ``ICRS`` and ``CIRS``.
This frame complements ``AltAz`` to give observed coordinates 
```

Comment (mkbrewer) [reply docs/changes/coordinates/11676.feature.rst]: I clicked on Commit suggestion. Does that actually make the change that you suggested?

Comment (mkbrewer) [reply docs/changes/coordinates/11676.feature.rst]: I guess it did.

**Comment by mhvk** [comment]
@StuartLittlefair - would you have some time to review this HADec implementation?

File: docs/changes/coordinates/11676.feature.rst
Comment (StuartLittlefair) [inline docs/changes/coordinates/11676.feature.rst:3]: can we clarify this what's new entry for users without more coordinates knowledge? Perhaps "...complements ``AltAz`` to give observed coordinates (hour angle and declination) in the ``ITRS`` for an equatorially mounted telescope."?

**Review by StuartLittlefair — APPROVED** [review]
Thanks so much @mkbrewer and @mhvk.

As you promised @mkbrewer the fix here is very straightforward and nicely implemented. I'm really sorry we didn't get round to implementing it, and that review took me so long. 

I've got no issues with the code here at all, but I think what we are missing is a single test that a conversion into the new HaDec frame gives the "right" answer. 

I don't think an explicit test of the HaDec results is necessary - the transform code is well tested by the `AltAz` tests and I think that any silly changes to the codebase would break the round-tripping that is tested. We'll just have to be careful to implement such a test if anyone ever splits the AltAz and HaDec transforms - perhaps if we ever move away from erfa, for example.

File: docs/changes/coordinates/11676.feature.rst
Comment (mkbrewer) [reply docs/changes/coordinates/11676.feature.rst:3]: Sure, no problem. Thanks for the review.

File: astropy/coordinates/tests/test_intermediate_transformations.py
Comment (mhvk) [inline astropy/coordinates/tests/test_intermediate_transformations.py:322]: @mkbrewer - to address @StuartLittlefair's larger comment, I'd suggest to here also check the numerical values of one of these three, just to ensure we do not later do something silly that makes, e.g., `HADec` be just an alias for `AltAz` (as is, I think all the tests would pass if one were to do that!).

Comment (mkbrewer) [reply astropy/coordinates/tests/test_intermediate_transformations.py:322]: OK. I'll go ahead and create a precision test.

**Comment by mhvk** [comment]
@mkbrewer - yes, with units those things should "Just work"! The changes look good, and I'll get this in once the tests pass.  Thanks again!

**Comment by mhvk** [comment]
Sadly the new test fails...

**Comment by mhvk** [comment]
OK, it is just a different wrapping of `ha`. I guess that brings up the question of whether HA should be between -12 and 12 hours, or between 0 and 24...

**Comment by mhvk** [comment]
OK, I pushed a fix, and also added a test that does not rely on remote data - using the "overhead" test so that the answers are known to be right by inspection. Will merge if tests now pass.

**Comment by mkbrewer** [comment]
@mhvk Hour angle always ranges from -12 to +12.

**Comment by mhvk** [comment]
Yes, I felt that made more sense, but that is not what your `HADec` class does....  I'll push an update.

**Comment by mkbrewer** [comment]
That's odd. Both `atioq()` and `ae2hd()` return the result of `atan2()`. This should always lie between `-pi` and `pi`.

**Comment by mhvk** [comment]
@mkbrewer - the problem is that the `[Unit]SphericalRepresentation` class uses a `Longitude` for which the default wrap angle is 360 degrees.  I now fixed it following what is done in `SkyOffsetFrame` - but I'll raise another issue that we should make this more flexible - see #11767.

Anyway, I think the current fix is the best we can do without a more general solution... And since tests pass, I'll merge it.  Thanks again!

**Pull request merged**
